### PR TITLE
PA-708: Add test for IssueDeleteOfIncrementalBackupsAndRestore

### PR DESCRIPTION
**Pull request by kshithijiyer-px**
**What this PR does / why we need it**:
- [x] Add new file `backup_delete_test.go`.
- [x] Move test` DeleteIncrementalBackupsAndRecreateNew` to `backup_delete_test.go`.
- [x] Add test for `IssueDeleteOfIncrementalBackupsAndRestore`.
- [x] Add `IsCloudCredPresent()` helper function. 
- [x] Change `CleanupCloudSettingsAndClusters()` to use `IsCloudCredPresent()`.
- [x] Increase timeout by 10 mins.  

Dashboard URL: 
http://aetos.pwx.purestorage.com/resultSet/testSetID/129923
~http://aetos.pwx.purestorage.com/resultSet/testSetID/125243~

**Which issue(s) this PR fixes** (optional)
Closes # PA-708

**Special notes for your reviewer**:
Will Squash and merge don't worry about the commits.

### CODE REVIEW

File: tests/backup/backup_delete_test.go [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+			log.InfoD("Taking backup of applications")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			// Full backup
+			for _, namespace := range bkpNamespaces {
+				fullBackupName = fmt.Sprintf("%s-%s-%v", "full-backup", namespace, time.Now().Unix())
+				backupNames = append(backupNames, fullBackupName)
+				err = CreateBackup(fullBackupName, SourceClusterName, customBackupLocationName, backupLocationUID, []string{namespace},
+					labelSelectors, orgID, clusterUid, "", "", "", "", ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying backup [%s] creation", fullBackupName))
```
Comment (kphalgun-px) [inline tests/backup/backup_delete_test.go:293]: How about fmt.Sprintf("Verifying full backup [%s] creation", fullBackupName)?

File: tests/backup/backup_delete_test.go [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+			}
+
+			// Incremental backup
+			for _, namespace := range bkpNamespaces {
+				incrementalBackupName = fmt.Sprintf("%s-%s-%v", "incremental-backup", namespace, time.Now().Unix())
+				incrementalBackupNames = append(incrementalBackupNames, incrementalBackupName)
+				err = CreateBackup(incrementalBackupName, SourceClusterName, customBackupLocationName, backupLocationUID, []string{namespace},
+					labelSelectors, orgID, clusterUid, "", "", "", "", ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying backup [%s] creation", incrementalBackupName))
```
Comment (kphalgun-px) [inline tests/backup/backup_delete_test.go:302]: How about fmt.Sprintf("Verifying incremental backup [%s] creation", incrementalBackupName)?

File: tests/backup/backup_delete_test.go [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+			for _, provider := range providers {
+				cloudCredUID = uuid.New()
+				cloudCredUidList = append(cloudCredUidList, cloudCredUID)
+				backupLocationUID = uuid.New()
+				credName := fmt.Sprintf("autogenerated-cred-%v", time.Now().Unix())
+				CreateCloudCredential(provider, credName, cloudCredUID, orgID)
+				log.InfoD("Created Cloud Credentials with name - %s", credName)
+				customBackupLocationName = fmt.Sprintf("autogenerated-backup-location-%v", time.Now().Unix())
+				backupLocationMap[backupLocationUID] = customBackupLocationName
+				err := CreateBackupLocation(provider, customBackupLocationName, backupLocationUID, credName, cloudCredUID, getGlobalBucketName(provider), orgID, "")
```
Comment (kphalgun-px) [inline tests/backup/backup_delete_test.go:79]: How about updating backupLocationMap after CreateBackupLocation? If CreateBackupLocation location fails, CleanupCloudSettingsAndClusters in JustAfterEach might throw an error saying backupLocation does not exist.

File: tests/backup/backup_delete_test.go [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+			log.InfoD("Creating cloud credentials and backup location")
+			for _, provider := range providers {
+				cloudCredUID = uuid.New()
+				cloudCredUidList = append(cloudCredUidList, cloudCredUID)
+				backupLocationUID = uuid.New()
+				credName = fmt.Sprintf("autogenerated-cred-%v", time.Now().Unix())
+				CreateCloudCredential(provider, credName, cloudCredUID, orgID)
+				log.InfoD("Created Cloud Credentials with name - %s", credName)
+				customBackupLocationName = fmt.Sprintf("autogenerated-backup-location-%v", time.Now().Unix())
+				backupLocationMap[backupLocationUID] = customBackupLocationName
```
Comment (kphalgun-px) [inline tests/backup/backup_delete_test.go:265]: How about updating backupLocationMap after CreateBackupLocation? If CreateBackupLocation location fails, CleanupCloudSettingsAndClusters in JustAfterEach might throw an error saying backupLocation does not exist.

File: tests/backup/backup_delete_test.go [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+			log.InfoD("Creating cloud credentials and backup location")
+			for _, provider := range providers {
+				cloudCredUID = uuid.New()
+				cloudCredUidList = append(cloudCredUidList, cloudCredUID)
+				backupLocationUID = uuid.New()
+				credName := fmt.Sprintf("autogenerated-cred-%v", time.Now().Unix())
+				CreateCloudCredential(provider, credName, cloudCredUID, orgID)
+				log.InfoD("Created Cloud Credentials with name - %s", credName)
+				customBackupLocationName = fmt.Sprintf("autogenerated-backup-location-%v", time.Now().Unix())
+				backupLocationMap[backupLocationUID] = customBackupLocationName
```
Comment (kphalgun-px) [inline tests/backup/backup_delete_test.go:78]: How about updating backupLocationMap after CreateBackupLocation? If CreateBackupLocation location fails, CleanupCloudSettingsAndClusters in JustAfterEach might throw an error saying backupLocation does not exist.

File: tests/backup/backup_delete_test.go [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+			log.FailOnError(err, "Fetching px-central-admin ctx failed")
+			for _, incrementalBackupName := range incrementalBackupNames2 {
+				bkpUid, err := backupDriver.GetBackupUID(ctx, incrementalBackupName, orgID)
+				log.FailOnError(err, "Unable to fetch backup UID - %s", incrementalBackupName)
+				bkpInspectReq := &api.BackupInspectRequest{
+					Name:  incrementalBackupName,
+					OrgId: orgID,
+					Uid:   bkpUid,
+				}
+				bkpInspectResponse, _ := backupDriver.InspectBackup(ctx, bkpInspectReq)
```
Comment (kphalgun-px) [inline tests/backup/backup_delete_test.go:156]: You might want to handle the error.

File: tests/backup/backup_delete_test.go (hunk 5 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_delete_test.go:78]: Done

File: tests/backup/backup_delete_test.go (hunk 4 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_delete_test.go:265]: Done

File: tests/backup/backup_delete_test.go (hunk 6 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_delete_test.go:156]: Done

File: tests/backup/backup_delete_test.go (hunk 1 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_delete_test.go:293]: Done

File: tests/backup/backup_delete_test.go (hunk 2 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_delete_test.go:302]: Done

**Review by kphalgun-px — APPROVED** [review]
LGTM

**Comment by mkoppal-px** [comment]
In the test `DeleteIncrementalBackupsAndRecreateNew`, please log the backupId to know exactly where it has an `incr` string or is it a full backup. Also print the backup name in the same line so that we know which backup name is in question.

**Comment by kshithijiyer-px** [comment]
> In the test `DeleteIncrementalBackupsAndRecreateNew`, please log the backupId to know exactly where it has an `incr` string or is it a full backup. Also print the backup name in the same line so that we know which backup name is in question.

Done

File: tests/backup/backup_helper.go [hunk 7]
```diff
@@ -1323,6 +1324,21 @@ func IsBackupLocationPresent(bkpLocation string, ctx context.Context, orgID stri
...
 
+// IsCloudCredPresent checks whether the Cloud Cred is present or not
+func IsCloudCredPresent(cloudCredName string, cloudCredUID string, ctx context.Context, orgID string) (bool, error) {
+	cloudCredInspectRequest := &api.CloudCredentialInspectRequest{
+		OrgId:          orgID,
+		Name:           cloudCredName,
+		IncludeSecrets: false,
+		Uid:            cloudCredUID,
+	}
+	_, err := Inst().Backup.InspectCloudCredential(ctx, cloudCredInspectRequest)
```
[the remark this answers is not in the document]
Comment (mkoppal-px) [reply tests/backup/backup_helper.go:1335]: I agree that it is more expensive. But I was thinking from a guaranteed result perspective. Also, checkout the `getAllBackupLocations` functions. I thought we can do it similar to that.

[the remark this answers is not in the document]
Comment (mkoppal-px) [reply tests/backup/backup_helper.go:1335]: But if you are sure that an error given by this function always means the cloud cred is not present anymore, we can go ahead with this approach. I don't have a strong view here about the approach.

**Comment by mkoppal-px** [comment]
Please help me understand this. We cannot delete `incrementalBackupNames` unless `incrementalBackupNames2` are deleted right?
I mean we have to delete backups in the descending order of creation (latest should be deleted first) right?

**Comment by mkoppal-px** [comment]
Also, I think you are taking 2 incremental backups and finally trying to restore the first full backup that was taken right?

File: tests/backup/backup_delete_test.go [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+					dash.VerifyFatal(strings.Contains(backupId, "incr"), true,
+						fmt.Sprintf("Check if the backup %s is incremental or not", incrementalBackupName))
+				}
+			}
+		})
+		Step("Restoring the backed up namespaces", func() {
+			log.InfoD("Restoring the backed up namespaces")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			for index, namespace := range bkpNamespaces {
```
Comment (mkoppal-px) [inline tests/backup/backup_delete_test.go:169]: Why can't we simply loop over `backupNames` and restore each backup? There is no need for us to use the namespace name in restore name. We would anyway know the namespace name from the name of the backup.

File: tests/backup/backup_delete_test.go [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+				restoreNames = append(restoreNames, restoreName)
+				log.InfoD("Restoring [%s] namespace from the [%s] backup", namespace, backupNames[index])
+				err = CreateRestore(restoreName, backupNames[index], namespaceMapping, destinationClusterName, orgID, ctx, make(map[string]string))
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Creating restore [%s]", restoreName))
+			}
+		})
+		Step("Validate applications", func() {
+			log.InfoD("Validate applications")
+			ValidateApplications(contexts)
+		})
```
Comment (mkoppal-px) [inline tests/backup/backup_delete_test.go:183]: Why do we need this? We would just be validating the apps deployed at the beginning of the test.

File: tests/backup/backup_delete_test.go [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+	. "github.com/onsi/ginkgo"
+	"github.com/pborman/uuid"
+	api "github.com/portworx/px-backup-api/pkg/apis/v1"
+	"github.com/portworx/torpedo/drivers/backup"
+	"github.com/portworx/torpedo/drivers/scheduler"
+	"github.com/portworx/torpedo/pkg/log"
+	. "github.com/portworx/torpedo/tests"
+)
+
+// IssueDeleteOfIncrementalBackupsAndRestore Issue delete of incremental backups in between and try to restore from
```
[the remark this answers is not in the document]
Comment (kshithijiyer-px) [reply tests/backup/backup_delete_test.go:17]: Done

File: tests/backup/backup_delete_test.go (hunk 8 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_delete_test.go:169]: Make sense changing it in the next PR.

File: tests/backup/backup_delete_test.go (hunk 9 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_delete_test.go:183]: Remove got accidentally  pasted twice.

File: tests/backup/backup_helper.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (kshithijiyer-px) [reply tests/backup/backup_helper.go:1335]: Okay changing it in the next patch

**Comment by kshithijiyer-px** [comment]
> Please help me understand this. We cannot delete `incrementalBackupNames` unless `incrementalBackupNames2` are deleted right? I mean we have to delete backups in the descending order of creation (latest should be deleted first) right?

So the idea here is that we need to have 3 backups 1 full and 2 incremental backups. Then issue delete of incremental  backup 1 and try to restore from incremental  backup 2. The backup incremental  backup 1 should go into delete pending state and restore should be successful.

**Comment by mkoppal-px** [comment]
The mentioned aetos dashboard link has a failed test. Can you also trigger a run with all the tests using BYOC job?

**Comment by kshithijiyer-px** [comment]
> The mentioned aetos dashboard link has a failed test. Can you also trigger a run with all the tests using BYOC job?

Done @mkoppal-px

File: tests/backup/backup_helper.go [hunk 11]
```diff
@@ -865,14 +865,19 @@ func CleanupCloudSettingsAndClusters(backupLocationMap map[string]string, credNa
...
 		}
+		err := DeleteCloudCredential(credName, orgID, cloudCredUID)
+		Inst().Dash.VerifySafely(err, nil, fmt.Sprintf("Verifying deletion of cloud cred [%s]", credName))
 		cloudCredDeleteStatus := func() (interface{}, bool, error) {
-			err := DeleteCloudCredential(credName, orgID, cloudCredUID)
-			if err != nil {
-				return "", true, fmt.Errorf("deleting cloud cred %s", credName)
+			status, err2 := IsCloudCredPresent(credName, ctx, orgID)
+			if err2 != nil {
+				return "", true, fmt.Errorf("deleting cloud cred %s still present with error %v", credName, err2)
```
Comment (snigdha-px) [inline tests/backup/backup_helper.go:873]: "deleting" word not required

File: tests/backup/backup_delete_test.go [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+				}
+				log.InfoD("Restoring %s backup", backupName)
+				err = CreateRestore(restoreName, backupName, namespaceMapping, destinationClusterName, orgID, ctx, make(map[string]string))
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Creating restore [%s]", restoreName))
+				restoreNames = append(restoreNames, restoreName)
+			}
+		})
+	})
+
+	JustAfterEach(func() {
```
[the remark this answers is not in the document]
Comment (kshithijiyer-px) [reply tests/backup/backup_delete_test.go:182]: Deleting the Backuplocation with true will take care of it.

[the remark this answers is not in the document]
Comment (snigdha-px) [reply tests/backup/backup_delete_test.go:182]: ok

File: tests/backup/backup_delete_test.go [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+	. "github.com/onsi/ginkgo"
+	"github.com/pborman/uuid"
+	api "github.com/portworx/px-backup-api/pkg/apis/v1"
+	"github.com/portworx/torpedo/drivers/backup"
+	"github.com/portworx/torpedo/drivers/scheduler"
+	"github.com/portworx/torpedo/pkg/log"
+	. "github.com/portworx/torpedo/tests"
+)
+
+// IssueDeleteOfIncrementalBackupsAndRestore  Issues delete of incremental backups in between and tries to restore from
```
[the remark this answers is not in the document]
Comment (kshithijiyer-px) [reply tests/backup/backup_delete_test.go:17]: Nice catch! Like your attention to details. Thanks.

File: tests/backup/backup_helper.go (hunk 11 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_helper.go:873]: Done.

**Comment by kshithijiyer-px** [comment]
Minor fixes in string no need to do any changes.

**Review by snigdha-px — APPROVED** [review]
LGTM

**Review by ak-px — APPROVED** [review]

File: tests/backup/backup_helper.go [hunk 14]
```diff
@@ -867,14 +867,19 @@ func CleanupCloudSettingsAndClusters(backupLocationMap map[string]string, credNa
 			_, err = task.DoRetryWithTimeout(backupLocationDeleteStatusCheck, cloudAccountDeleteTimeout, cloudAccountDeleteRetryTime)
 			Inst().Dash.VerifySafely(err, nil, fmt.Sprintf("Verifying backup location deletion status %s", bkpLocationName))
 		}
+		err := DeleteCloudCredential(credName, orgID, cloudCredUID)
+		Inst().Dash.VerifySafely(err, nil, fmt.Sprintf("Verifying deletion of cloud cred [%s]", credName))
 		cloudCredDeleteStatus := func() (interface{}, bool, error) {
-			err := DeleteCloudCredential(credName, orgID, cloudCredUID)
-			if err != nil {
-				return "", true, fmt.Errorf("deleting cloud cred %s", credName)
+			status, err2 := IsCloudCredPresent(credName, ctx, orgID)
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:873]: Declare `var err error` in the function and keep using err everywhere

File: tests/backup/backup_helper.go [hunk 15]
```diff
@@ -1325,6 +1330,26 @@ func IsBackupLocationPresent(bkpLocation string, ctx context.Context, orgID stri
...
+	if err != nil {
+		return false, err
+	}
+	for _, cloudCredObj := range cloudCredObjs.GetCloudCredentials() {
+		if cloudCredObj.GetName() == cloudCredName {
+			log.Infof("Cloud Credential [%s] is present", cloudCredName)
+			return true, nil
+		}
+	}
+	log.Infof("Cloud Cred Object %s", cloudCredName)
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:1349]: Remove this log entry

File: tests/backup/backup_delete_test.go [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+			for _, namespace := range bkpNamespaces {
+				incrementalBackupName = fmt.Sprintf("%s-%s-%v", "incremental-backup", namespace, time.Now().Unix())
+				incrementalBackupNames2 = append(incrementalBackupNames2, incrementalBackupName)
+				err = CreateBackup(incrementalBackupName, SourceClusterName, customBackupLocationName, backupLocationUID, []string{namespace},
+					labelSelectors, orgID, clusterUid, "", "", "", "", ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying incremental backup [%s] creation", incrementalBackupName))
+			}
+			log.Infof("List of backups - %v", backupNames)
+			log.Infof("List of Incremental backups Set 1 - %v", incrementalBackupNames)
+			log.Infof("List of Incremental backups Set 2 - %v", incrementalBackupNames2)
```
Comment (mkoppal-px) [inline tests/backup/backup_delete_test.go:128]: Would it be better we push this to Aetos dashboard with `log.InfoD` ?

File: tests/backup/backup_helper.go (hunk 15 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_helper.go:1349]: Done

File: tests/backup/backup_helper.go (hunk 14 above)
Comment (kshithijiyer-px) [reply tests/backup/backup_helper.go:873]: Done

**Review by mkoppal-px — APPROVED** [review]
LGTM

**Pull request merged**